### PR TITLE
PXF filter pushdown

### PR DESCRIPTION
**Pull request by leskin-in**
Implement filter pushdown for external sources accessible by PXF protocol.

Pushdown can be implemented for other `custom` protocols, too, and the changes will reside in the code of GPDB extension for that protocol, not in GPDB core.

The feature is disabled by default and can be enabled by setting a GUC parameter `enable_filter_pushdown` to `on`.

#### Principle

The filter quals are extracted from a query in [`ExternalNext()`](https://github.com/arenadata/db/blob/filter_pushdown_pxf/src/backend/executor/nodeExternalscan.c#L129), passed (as a `List*`) to `url_custom_fopen()`, and pushed into `ExtProtocolData` located in [`URL_CUSTOM_FILE`](https://github.com/arenadata/db/blob/filter_pushdown_pxf/src/backend/access/external/url_custom.c#L27). The latter is then provided to an accessor of external data source (in case of PXF the entrypoint is [`pxfprotocol_import()`](https://github.com/arenadata/db/blob/filter_pushdown_pxf/gpAux/extensions/pxf/src/pxfprotocol.c#L124)).

The external data source accessor retrieves the constraints from `ExtProtocolData` and processes them properly (in case of PXF the processing is a custom constraints serialization and encoding; the code is in `pxffilters.c` (the file comes from Apache HAWQ, with minor modifications), and the rules can be found in [`pxf_serialize_filter_list()`](https://github.com/arenadata/db/blob/filter_pushdown_pxf/gpAux/extensions/pxf/src/pxffilters.c#L387)).

Finally, the accessor makes a request to the external data source with the constraints from GPDB included.

When the data is returned to GPDB, the same constraints are applied to the result again. This keeps the output consistent in case constraints were not processed properly by PXF (or some other) extension or PXF (or other external data source) was unable to process them properly.

#### PXF accessor

PXF external source accessor has certain limitations. When they are not met, no constraints will be passed to PXF. The limitations are as follows:
* Multiple constraints must be `AND`ed (`OR` is not allowed);
* Only `>`, `<`, `=`, `<>`, `BETWEEN`, `IN` and `LIKE` constraint operators are supported;
* Only `INT`, `FLOAT`, `REAL`, `TEXT`, `CHAR`, `DATE` and `TIMESTAMP` constraint data types are supported. All constraints (except for those of type `INT`) should be explicitly cast to the type of field they are applied to;

If any of the limitations is not met, a query will be executed properly, but pushdown will be disabled for it.

Also, every PXF plugin has its own set of supported operators and data types for pushdown, thus, even if the constraints were properly processed by GPDB PXF extension, PXF may not pass them to the external data source.

### CODE REVIEW

**Comment by ivannovick** [comment]
@kongyew @macroyuyang can you review this please?

**Comment by Eulerizeit** [comment]
@leskin-in is this intended for master only or do you wish for it to be backported to 5.X?

**Comment by kapustor** [comment]
Hi @Eulerizeit 

Me and @leskin-in are from one team. Yes, we will be glad to see this feature in current 5.X version.
Actualy, we were developing it in 5X_stable branch in our fork, but, as I understand, we can make PR only to master branch with following backport.

Thanks.

**Comment by leskin-in** [comment]
@Eulerizeit, the code for `5X_STABLE` has almost no differences from the one for `master` branch, and we have already prepared it (except for the last changes in tests). See https://github.com/arenadata/gpdb/tree/5X_filter_pushdown_pxf.

**Comment by kongyew** [comment]
@denalex  @shivzone @lavjain  Please review and comment on this PR.

**Comment by Eulerizeit** [comment]
> Hi @Eulerizeit
>
>Me and @leskin-in are from one team. Yes, we will be glad to see this feature in current 5.X version.
Actualy, we were developing it in 5X_stable branch in our fork, but, as I understand, we can make PR only to master branch with following backport.

You are absolutely following the correct process. The reason I was asking is because if this were  for the master branch only I was going to propose holding off since we are getting the stable FDW API soon and that will, I hope, cause a fair amount of churn in this code. Since that won't be backported to 5.x then having this PR go through the master then backport process totally makes sense. 

thanks for the PR.
Rob

**Comment by macroyuyang** [comment]
@pengzhout @pf-qiu Please have a look at this.

**Comment by macroyuyang** [comment]
I am wondering whether the compare logic is the same, for in GPDB compared to that in external source? Maybe locale is different. And there will be a collation feature support in GPDB as well.

**Comment by kapustor** [comment]
Hi @macroyuyang 
We have tested the pushdown feature with JDBC plugin with the following data sources:
Oracle DB, mysql, MSSQL, PostgreSQL, Apache Ignite.
For example, here is a comparison compatibility matrix for Oracle DB:
image
(The red cell means that using this comparison causes the pushdown to be turned off and the query runs as usual)
As all of these databases are following ANSI SQL standard, their compare logic (as I hope) is the same. However, there might be some differences in data types accuracy: for example, MSSQL by default stores three digits after second (milliseconds), while Oracle DB stores no digits after second. Thus, filter in GP like `where date1 ='2017-10-01 12:18:23.123'` can return something from MSSQL but will return nothing from Oracle DB, though it can contain a row with `date1 ='2017-10-01 12:18:23'`. As we think, users must care about external data types by themselves.
Thanks!

**Comment by shivzone** [comment]
Functionally looks good.
Need to add unit tests to cover positive and negative tests of handling logical and comparison operators as well as the various data types. Want to make sure we don't pass through filters/data types that aren't handled on the pxf server side.

**Comment by macroyuyang** [comment]
Thanks, @kapustor.
+1 for more test cases.
From the code, whether we push down filters to external source (like S3) is determined by Java side?

**Comment by leskin-in** [comment]
The unit tests for PXF come from Apache HAWQ (with minor modifications due to the changes in the list of supported operators). However, a segfault happens when any postgres memory allocation function (for example, `palloc`) is called.

Could anyone explain what is the cause of that behaviour and how to deal with it?

**Comment by macroyuyang** [comment]
The palloc is a wrapper of `MemoryContextAlloc(CurrentMemoryContext, (sz))`. So the global variable CurrentMemoryContext should be set properly or you could mock palloc?

**Comment by shivzone** [comment]
@leskin-in I do agree the PXF plugin would get the final say when it comes to dealing with the filters. That said, you have used the following gating functions - 1. `supported_operator_type_op_expr` for checking the operators and 2. `supported_filter_type()` for checking the data type, prior to passing the filters to the PXF service.

**Comment by shivzone** [comment]
@hlinnaka @danielgustafsson can you guys please chime in on this PR

**Comment by danielgustafsson** [comment]
Skimming the diff against the core parts, nothing really sticks out (in a non-FDW world for external tables). I didn't look closely at the PXF parts as thats well outside my area of expertise.

**Comment by macroyuyang** [comment]
Ideally, any calls going to Greenplum in unittest should be mocked. StartupXLOG has different functionalities in different mode. In normal, it will return after redo all XLOGRecords, but in recovery mode, it never return, and will request postmaster to start WalReceiver process, which is more complicated

**Comment by leskin-in** [comment]
I wonder why does the last commit make tests fail.

Now I cannot access the test results from Pivotal CI (for any build). Could anyone explain the reason? Perhaps it is somehow connected with authentication, as I can see an error `Failed to load resource: the server responded with a status of 401 ()` for the address https://gpdb.bosh.pivotalci.info/auth/userinfo in Chromium.

**Comment by kapustor** [comment]
@danielgustafsson @hsyuan Hi guys!
What else can we do to help this PR merged?

**Comment by danielgustafsson** [comment]
I only have one objection which I consider a :-1: for merging until addressed: the `enable_filter_pushdown` GUC is too broadly named. As we merge with upstream we will get qual and join pushdown into foreign data wrappers and this GUC will then be confusing. I think we need to get "external tables" into the GUC in some way to convey meaning. Maybe simply `external_table_filter_pushdown`? Since the GUC is boolean we don't really need "enable" in the name IMHO.

**Comment by danielgustafsson** [comment]
Also, I assume the patchset will be squashed into a single commit before pushing?

**Comment by leskin-in** [comment]
@danielgustafsson Shall I squash all the current commits? As far as I understand (am I making a mistake?), the squash is done by the person who actually merges the PR.

**Comment by danielgustafsson** [comment]
> On 6 Jun 2018, at 22:12, Ivan Leskin <notifications@github.com> wrote:
> 
> @danielgustafsson Shall I squash all the current commits? As far as I understand (am I making a mistake?), the squash is done by the person who actually merges the PR.

Either way is fine. I was under the impression that you had commit access, sorry for the confusion.

File: src/backend/utils/misc/guc_gp.c [hunk 1]
```diff
@@ -480,7 +480,7 @@ char	   *gp_default_storage_options = NULL;
 
 int			writable_external_table_bufsize = 64;
 
-bool		enable_filter_pushdown = false;
+bool		gp_external_enable_filter_pushdown = false;
```
Comment (danielgustafsson) [inline src/backend/utils/misc/guc_gp.c]: +1 on this rename

**Comment by leskin-in** [comment]
@danielgustafsson,
Is there something that should be done before the PR can be merged? (You have not given your approval on it).

**Comment by danielgustafsson** [comment]
I gave a +1 on the GUC rename which was the only reservation I had, so I'm good for merging this. I am however not a maintainer of the PXF code, and far from an expert on it, so I'm waiting for a PXF maintainer to do the actual merge.

**Comment by denalex** [comment]
Thank you @danielgustafsson for approving the changes on the core GPDB part. The PXF committers will review the PXF portion next week. Once there are no more code change suggestions, we will have to run this PR through GPDB/PXF regression suite, where we will need to enable currently disabled tests that exercise variety of predicate pushdown scenarios. Once the integration tests are green we will merge the PR. Will update on the progress of this effort in a few days.

**Comment by shivzone** [comment]
I still see a few references to "HAWQ" in the types and operators supported in pxffilters.c and .h files.
Please modify/remove them

**Comment by leskin-in** [comment]
@shivzone,
The necessary corrections have been made. If not all references to `HAWQ` you've noted were renamed, please, let me know.

**Comment by ivannovick** [comment]
@yydzero would you please take a look at this also?

**Comment by yydzero** [comment]
+1 for @danielgustafsson from core side.

@denalex any updates from PXF side? 

Thanks for contribution @leskin-in @kapustor

**Comment by benchristel** [comment]
Verified the predicate pushdown functionality by running the internal regression suite. Everything looks good so far.

**Comment by ashwinstar** [comment]
On Fri, Jun 22, 2018 at 11:15 AM Ben Christel <notifications@github.com>
wrote:

> Verified the predicate pushdown functionality by running the internal
> regression suite. Everything looks good so far.
>
Great, lets merge this in.

**Comment by Eulerizeit** [comment]
Awesome to see this in. Thanks for the contribution @leskin-in

**Comment by ivannovick** [comment]
+1 great work all!

IMO this is great to see a COMMUNITY of members all benefiting from GPDB and contributing to it.

Love it

**Comment by ivannovick** [comment]
@kapustor @leskin-in can you provide some content you suggest be added to the official documentation for  this new capability and perhaps @dyozie can help to get the new capability added to our docs (assuming its not already).
Thx!

**Comment by kapustor** [comment]
Awesome to see the merge!
Thank you all guys for your time, comments and participation. I hope this feature will make Greenplum one step closer to the perfect MPP DB.
Happy to be part of GP OS community.
@ivannovick Yep, we will provide some basic docs texts during this week and prepare backport to 5X_stable branch

**Comment by lavjain** [comment]
@kapustor We already backported the PR to 5X_STABLE

**Comment by leskin-in** [comment]
@ivannovick ,
Basic documentation (a reference in the GUC list and a shallow description of pushdown itself) was included in the PR by [this commit](https://github.com/greenplum-db/gpdb/commit/0bf60e47f29be61c0763bb8497df4fb203d17b1a#diff-4600fda9b3a17a590558ff97fbeecc06R3676).

For sure, it's worth adding extra information on how pushdown (for any extensions) works, what operators and types are supported by PXF extension and maybe something else, too. I think @kapustor and me will be able to propose the exact changes before the end of the upcoming week.

Shall we create a new PR for that?

**Comment by ivannovick** [comment]
@dyozie have you reviewed and need any more info for documentation?

**Comment by leskin-in** [comment]
I have added some references to the pushdown feature and described what data types are supported by GPDB PXF extension; see #5193.

**Pull request merged**